### PR TITLE
Reduce usage of local cells

### LINKED ISSUE #2899: Reduce uses of the local_cells global
Many methods directly use the `local_cells` global, in most of the cases to iterate over the particles.
This exposes implementation details of the cell system to the methods, which isn't actually needed.
For those methods the `ParticleRange` to be iterated over should be passed as an argument.

### PR DESCRIPTION
**Pull request by KaiSzuttor**
Fixes part of #2899

Description of changes:
 - replacement for #2991

### CODE REVIEW

**Review by fweik — DISMISSED** [review]

File: src/core/forces.hpp [hunk 1]
```diff
@@ -60,10 +62,11 @@ void init_forces_ghosts();
  *  <li> Calculate long range interaction forces
  *  </ol>
  */
-void force_calc();
+void force_calc(CellStructure &cellStructure);
```
Comment (fweik) [inline src/core/forces.hpp]: Naming convention

File: src/core/cells.cpp [hunk 2]
```diff
@@ -68,6 +68,15 @@ double max_range = 0.0;
 unsigned resort_particles = Cells::RESORT_NONE;
 int rebuild_verletlist = 1;
 
+
+ParticleRange CellStructure::get_local_particles() const {
```
Comment (fweik) [inline src/core/cells.cpp]: Please add instead methods that return the cell lists, and get the particles from them. Also I am
not in favor of the `get_`, the verb suggests an action, but this only refers to an existing entity.

File: src/core/cells.cpp [hunk 3]
```diff
@@ -68,6 +68,15 @@ double max_range = 0.0;
...
+  return local_cells.particles();
+}
+
+ParticleRange CellStructure::get_ghost_particles() const {
```
Comment (fweik) [inline src/core/cells.cpp]: see `get_local_particles`

File: src/core/communication.cpp [hunk 4]
```diff
@@ -410,7 +410,7 @@ void mpi_gather_stats(int job, void *result, void *result_t, void *result_nb,
     break;
   case 4:
     mpi_call(mpi_gather_stats_slave, -1, 4);
-    predict_momentum_particles((double *)result);
+    predict_momentum_particles((double *)result, local_cells.particles());
```
Comment (fweik) [inline src/core/communication.cpp]: Get from `cell_structure`. In the end the `{local, ghost}_cells` will be removed.

File: src/core/communication.cpp [hunk 5]
```diff
@@ -450,7 +450,7 @@ void mpi_gather_stats_slave(int, int job) {
     pressure_calc(nullptr, nullptr, nullptr, nullptr, 1);
     break;
   case 4:
-    predict_momentum_particles(nullptr);
+    predict_momentum_particles(nullptr, local_cells.particles());
```
Comment (fweik) [inline src/core/communication.cpp]: Get from `cell_structure`

File: src/core/communication.cpp [hunk 6]
```diff
@@ -661,7 +661,7 @@ void mpi_bcast_max_mu() {
 
 /***** GALILEI TRANSFORM AND ASSOCIATED FUNCTIONS ****/
 void mpi_kill_particle_motion_slave(int rotation) {
-  local_kill_particle_motion(rotation);
+  local_kill_particle_motion(rotation, local_cells.particles());
```
Comment (fweik) [inline src/core/communication.cpp]: Get from `cell_structure`

File: src/core/communication.cpp [hunk 7]
```diff
@@ -672,7 +672,7 @@ void mpi_kill_particle_motion(int rotation) {
 }
 
 void mpi_kill_particle_forces_slave(int torque) {
-  local_kill_particle_forces(torque);
+  local_kill_particle_forces(torque, local_cells.particles());
```
Comment (fweik) [inline src/core/communication.cpp]: Get from `cell_structure`

File: src/core/forces.cpp [hunk 8]
```diff
@@ -74,26 +74,27 @@ void init_forces() {
...
   }
 
-  for (auto &p : ghost_cells.particles()) {
+  for (auto &p : ghost_particles) {
     check_particle_force(&p);
   }
 }
 
-void force_calc() {
+void force_calc(CellStructure &cellStructure) {
```
Comment (fweik) [inline src/core/forces.cpp]: Should be `cell_structure` by our naming conventions.

File: src/script_interface/constraints/ShapeBasedConstraint.hpp [hunk 9]
```diff
@@ -61,7 +61,7 @@ class ShapeBasedConstraint : public Constraint {
       return shape_based_constraint()->total_force();
     }
     if (name == "min_dist") {
-      return shape_based_constraint()->min_dist();
+      return shape_based_constraint()->min_dist(local_cells.particles());
```
Comment (fweik) [inline src/script_interface/constraints/ShapeBasedConstraint.hpp]: Get from `cell_structure`

File: src/script_interface/mpiio/si_mpiio.hpp [hunk 10]
```diff
@@ -52,7 +53,7 @@ class MPIIOScript : public AutoParameters<MPIIOScript> {
                  field_value(bond, Mpiio::MPIIO_OUT_BND);
 
     if (name == "write")
-      Mpiio::mpi_mpiio_common_write(pref.c_str(), v);
+      Mpiio::mpi_mpiio_common_write(pref.c_str(), v, local_cells.particles());
```
Comment (fweik) [inline src/script_interface/mpiio/si_mpiio.hpp]: Get from `cell_structure`

File: src/core/integrate.cpp [hunk 11]
```diff
@@ -229,16 +229,18 @@ void integrate_vv(int n_steps, int reuse_forces) {
     // Communication step: distribute ghost positions
     cells_update_ghosts();
 
+    auto particles = local_cells.particles();
```
Comment (fweik) [inline src/core/integrate.cpp]: Get from `cell_structure`

File: src/core/integrate.cpp [hunk 12]
```diff
@@ -267,9 +269,11 @@ void integrate_vv(int n_steps, int reuse_forces) {
     ESPRESSO_PROFILER_CXX_MARK_LOOP_ITERATION(integration_loop, step);
     INTEG_TRACE(fprintf(stderr, "%d: STEP %d\n", this_node, step));
 
+    auto particles = local_cells.particles();
```
Comment (fweik) [inline src/core/integrate.cpp]: Get from `cell_structure`

File: src/core/integrate.cpp [hunk 13]
```diff
@@ -301,7 +305,7 @@ void integrate_vv(int n_steps, int reuse_forces) {
     if (n_rigidbonds) {
       cells_update_ghosts();
 
-      correct_pos_shake();
+      correct_pos_shake(local_cells.particles());
```
Comment (fweik) [inline src/core/integrate.cpp]: Get from `cell_structure`

File: src/core/integrate.cpp [hunk 14]
```diff
@@ -326,7 +330,9 @@ void integrate_vv(int n_steps, int reuse_forces) {
     // Propagate langevin philox rng counter
     langevin_rng_counter_increment();
 
-    force_calc();
+    particles = local_cells.particles();
```
Comment (fweik) [inline src/core/integrate.cpp]: Get from `cell_structure`

**Review by fweik — CHANGES_REQUESTED** [review]

File: src/core/rotate_system.cpp [hunk 15]
```diff
@@ -72,7 +73,7 @@ void mpi_rotate_system_slave(int, int) {
   std::array<double, 3> params;
   mpi::broadcast(comm_cart, params, 0);
 
-  local_rotate_system(params[0], params[1], params[2]);
+  local_rotate_system(params[0], params[1], params[2], local_cells.particles());
```
Comment (fweik) [inline src/core/rotate_system.cpp]: Sorry, missed these two: get from `cell_structure`

File: src/core/rotate_system.cpp [hunk 16]
```diff
@@ -81,7 +82,7 @@ void mpi_rotate_system(double phi, double theta, double alpha) {
   std::array<double, 3> params{{phi, theta, alpha}};
   mpi::broadcast(comm_cart, params, 0);
 
-  local_rotate_system(params[0], params[1], params[2]);
+  local_rotate_system(params[0], params[1], params[2], local_cells.particles());
```
Comment (fweik) [inline src/core/rotate_system.cpp]: Get from `cell_structure`

**Review by fweik — CHANGES_REQUESTED** [review]

File: src/core/cells.hpp [hunk 17]
```diff
@@ -147,6 +147,11 @@ struct CellStructure {
 
   bool use_verlet_list = true;
 
+  /** returns the global local_cells.particles() */
```
Comment (fweik) [inline src/core/cells.hpp]: The comments no longer are valid.

**Review by fweik — CHANGES_REQUESTED** [review]

**Comment by fweik** [comment]
Ok, I think this is a clear improvement an can be merged if the tests pass. @KaiSzuttor what do you think?

**Review by fweik — APPROVED** [review]

**Review by KaiSzuttor — COMMENTED** [review]
yes, I think it looks good now, just cannot approve my own PR

**Pull request merged**